// besst-usb-sdk/src/machineInformation/uart/battery.js
/**
 * Machine information UART battery
 * @module machineInformation/uart/battery
 */

const client = require('../../usbClient/client');
const common = require('../../usbClient/common');

/**
 * @function
 * @returns {Promise}
 */
module.exports.softwareVersionNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A1651010068000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.hardwareVersionNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A165301006A000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.modelNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A1650010067000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.serialNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A1652010069000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.tempreature = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A160801001F000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.voltage = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A1609010020000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.current = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A160A010021000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.relativeCapacity = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A160D010024000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.absoluteCapacity = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A160E010025000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.capacityLeft = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A160F010026000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.fullCapacity = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A1610010027000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.cycleCount = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A161701002E000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.cells1to7 = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A162401003B000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.cells8to14 = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A162501003C000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.currentChargeInterval = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A164701005E000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.maxChargeInterval = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '3A164801005F000D0A');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.setModeIn = () => {
    return client.batterySetModeIn();
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.setModeOut = () => {
    return client.batterySetModeOut();
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.update = (newValue) => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, newValue);
};


//////////////////
// WEBPACK FOOTER
// ./node_modules/besst-usb-sdk/src/machineInformation/uart/battery.js
// module id = 1079
// module chunks = 73
