// besst-usb-sdk/src/machineInformation/uart/ctl.js
/**
 * Machine information UART CTL
 * @module machineInformation/uart/ctl
 */

const client = require('../../usbClient/client');
const common = require('../../usbClient/common');

/**
 * @function
 * @returns {Promise}
 */
module.exports.softwareVersionNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '1150');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.hardwareVersionNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '1417');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.modelNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '1416');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.serialNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '1414');

};

/**
 * @function
 * @returns {Promise}
 */
module.exports.basicInfo = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '11512580f6');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.parameters = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, '1152');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.update = (newValue) => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_CTL, newValue);
};



//////////////////
// WEBPACK FOOTER
// ./node_modules/besst-usb-sdk/src/machineInformation/uart/ctl.js
// module id = 1080
// module chunks = 73
