// besst-usb-sdk/src/machineInformation/uart/hmi.js
/**
 * Machine information UART HMI
 * @module machineInformation/uart/hmi
 */

const client = require('../../usbClient/client');
const common = require('../../usbClient/common');

/**
 * @function
 * @returns {Promise}
 */
module.exports.softwareVersionNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '1210');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.hardwareVersionNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '1217');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.modelNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '110A1B');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.serialNumber = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '1214');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.serviceMileage = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '1216');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.totalGear = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '1218');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.totalMileage = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '1219');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.speedAndWheel = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '110D1E');
};

/**
 * @function
 * @returns {Promise}
 */
module.exports.clientCode = () => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, '110B1C');
};

/**
 * @function
 * @param {string} newValue - new value
 * @returns {Promise}
 */
module.exports.update = (newValue) => {
    return client.request_getUart(common.USB_CMD.USB_CMD_PC_TO_HMI, newValue);
};


//////////////////
// WEBPACK FOOTER
// ./node_modules/besst-usb-sdk/src/machineInformation/uart/hmi.js
// module id = 1081
// module chunks = 73
